backend/app/http_sig.py

```diff
@@ -0,0 +1,74 @@
+import base64
+import hashlib
+import json
+
+import http_sfv
+from app.settings import conf
+from http_message_signatures import (
+    HTTPMessageSigner,
+    HTTPMessageVerifier,
+    HTTPSignatureKeyResolver,
+    algorithms,
+)
+from jwt.jwks_client import PyJWKClient
+
+jwks_client = PyJWKClient(conf.HTTP_SIG_VERIFY_JWKS_URI)
+
+
+class ContentDigestMismatch(Exception):
+    pass
+
+
+class MissingContentDigest(Exception):
+    pass
+
+
+class AppKeyResolver(HTTPSignatureKeyResolver):
+    def __init__(self):
+        self.keys = {conf.PRIVATE_KEY.kid: conf.PRIVATE_KEY}
+
+    def resolve_private_key(self, key_id: str):
+        return self.keys[key_id].key
+
+    def resolve_public_key(self, key_id: str):
+        print(
+            f"Resolving public RSA key ({key_id}) from {conf.HTTP_SIG_VERIFY_JWKS_URI}"
+        )
+        return jwks_client.get_signing_key(key_id).key
+
+
+def inject_digest(headers, body):
+    digest = http_sfv.Dictionary({"sha-256": hashlib.sha256(body).digest()})
+    headers["Content-Digest"] = str(digest)
+
+
+def verify_content_digest(headers, body):
+    try:
+        content_digest = headers["Content-Digest"]
+    except KeyError:
+        raise MissingContentDigest
+    incoming_digest = content_digest.split(":")[1]
+    body_bytes = bytes(json.dumps(body), "utf8")
```

Comment:
This looks like it could easily fail if the productizer for any reason decides to do any kind of change to whitespace, order of keys for any reason aren't preserved correctly when loading and dumping the json, non-ascii characters are differently encoded etc. Would be better to pass the raw body into the function than dumping it again and hoping to get the same byte representation as was sent.